### PR TITLE
ARROW-10481: [R] Bindings to add, remove, replace Table columns 

### PR DESCRIPTION
**Pull request by jonkeane**
This allows for R users to use either `[[` or `$` syntax for removing, adding, or replacing table columns. 

If one supplies an index larger than the number of columns (e.g. `mtcars[[12]] <- vector`), the column is added to the end, the index is cast to a string (here: `"12"`) for the name. This might be better as an error instead of a silent cast for the name.

### CODE REVIEW

File: r/R/table.R
Comment (nealrichardson) [inline r/R/table.R:265]: I think we need to do the `- 1L` even if i is not character, otherwise we'll be off by one with an integer `i`.

File: r/R/table.R
Comment (nealrichardson) [inline r/R/table.R:270]: This is a trivial difference but it does save some work because `ncol == function (x) dim(x)[2L]`

```suggestion
      if (i <= x$num_columns) {
```

File: r/R/table.R
Comment (nealrichardson) [inline r/R/table.R:271]:
```suggestion
        i <- names(x)[i]
```

File: r/R/table.R
Comment (nealrichardson) [inline r/R/table.R:296]: Can you add a comment explaining why this `if` is here?

File: r/tests/testthat/test-Table.R
Comment (nealrichardson) [inline r/tests/testthat/test-Table.R:164]: Part of the behavior we want to test is the position of the new column, so let's assert the whole data.frame

```suggestion
  expect_data_frame(tab, cbind(tbl, new = letters[10:1]))
```

File: r/tests/testthat/test-Table.R
Comment (nealrichardson) [inline r/tests/testthat/test-Table.R:186]: Other tests we might want to add (and implement as necessary):

* `tab[[NA]] <- something`
* `tab[[NULL]] <- something`
* R input recycling (making sure we're consistent with how [[<-.data.frame works)
* assigning in an Arrow Array or ChunkedArray. `chunked_array()` should handle that but we should confirm
* any others?

**Review by nealrichardson — CHANGES_REQUESTED** [review]
Thanks! A few notes

File: r/R/table.R
Comment (jonkeane) [reply r/R/table.R:270]: I went looking for a function call to `num_columns()` and should have looked for an active binding, oops!

**Comment by jonkeane** [comment]
I added the extra tests and some code to get that working. 

I also started down the path of `[<-.Table` though there are more decisions that are under that rock (what to do about row manipulation, etc.), so left it be for now. It currently errors in R (though we probably want to write a more helpful message as part of this PR even if we kick that working to another branch/PR).

File: r/tests/testthat/test-Table.R
Comment (nealrichardson) [inline r/tests/testthat/test-Table.R:212]: I'm not sure that's the right error to raise there so maybe there should be better validation here.

Would also be important to test `NA_integer_` and `NA_real_` because those will not get rejected by the R->cpp type mapping.

File: r/tests/testthat/test-Table.R
Comment (nealrichardson) [inline r/tests/testthat/test-Table.R:213]:
```suggestion
  expect_error(tab[[NULL]] <- letters[10:1])
```

Since that is an error message coming from base R (right?), we shouldn't assert its exact message because it may be translated to local language (i.e. it only will pass if you're in an English locale).

File: r/R/table.R
Comment (nealrichardson) [inline r/R/table.R:277]: 1) We should probably only do recycling if `value` is not an ArrowObject (perhaps with a TODO to support Arrow Scalar recycling)
2) https://vctrs.r-lib.org/reference/vec_recycle.html may be a good solution instead of this bespoke code (we already import vctrs so it's available), or else use `base::rep_len` (cf. https://github.com/apache/arrow/blob/master/r/R/array.R#L291)

**Review by nealrichardson — CHANGES_REQUESTED** [review]
Thanks for pushing this ahead. 

Re: [<- method, I think a followup JIRA is acceptable. The current error message isn't awesome but is clear that it isn't implemented:

```
Error in tab[1, 1] <- 5 : object of type 'environment' is not subsettable
```

The other thing we should make a followup for is for doing the same thing for RecordBatch. _As it turns out_, that will require a little more C++ work because there is AddColumn and RemoveColumn but no SetColumn in record_batch.h.

File: r/tests/testthat/test-Table.R
Comment (jonkeane) [reply r/tests/testthat/test-Table.R:213]: Yeah, I added error checking into the method call for `[[<-` to match the behavior + catch `NA_integer_`. It looks like the NA checking is actually coming from cpp `arrow::r::Input<R_xlen_t>::type`, if I'm reading correctly. I put a TODO note to see about unifying / centralizing that checking

File: r/R/table.R
Comment (jonkeane) [reply r/R/table.R:277]: TIL about `vctrs::vec_recycle` that's really nice.

**Comment by jonkeane** [comment]
Ok, I opened https://issues.apache.org/jira/browse/ARROW-10520 and https://issues.apache.org/jira/browse/ARROW-10521 for the two follow ons, added some more NA and type checking, and replaced the value recycling with {{vctrs::vec_recycle}}

File: r/tests/testthat/test-Table.R
Comment (nealrichardson) [reply r/tests/testthat/test-Table.R]: Oh nice, I've been bitten by that before, so I guess either I was using `int` instead of `R_xlen_t` or `Rcpp` wasn't as helpful. TIL :)

File: r/R/table.R
Comment (nealrichardson) [inline r/R/table.R:267]: If we get a useful error message from the cpp layer, great, use that. Let the static(ish) typing do the work.

**Review by nealrichardson — COMMENTED** [review]
One note on the TODO but otherwise I think we're good enough for now.

Let's hold this for today see if Romain can merge #8256 tomorrow--I've made enough merge conflicts for him on that PR by now. If it's still not ready then, we can merge.

File: r/R/table.R
Comment (jonkeane) [reply r/R/table.R:267]: It's a mixture, for `NA_integer_` and `NA_real_` the error is the previous "missing value where TRUE/FALSE needed" with `NA_character_` though it removes/creates/replaced a column with the name ``NA``. So I would vote to keep the special casing for now and try catching in cpp or with a helper function we can use across `[[` and `[` methods later.

Comment (nealrichardson) [reply r/R/table.R:273]: I suspect that error comes from https://github.com/apache/arrow/pull/8579/files#diff-242f7f20fb4999f3bbe719d6977d2580231b89334326b588d46e9236c2e28fcdR285, i.e. before we hit cpp.

**Review by nealrichardson — APPROVED** [review]
I've rebased; will merge when CI is green

**Pull request closed without merging**
